### PR TITLE
TextView extensions

### PR DESCRIPTION
**Pull request by jaredsburrows**
Help add extensions and tests for https://github.com/android/android-ktx/issues/227 @JakeWharton @romainguy

### CODE REVIEW

**Comment by dovahkiin98** [comment]
all these extensions are already present in PR #269

**Comment by illuzor** [comment]
I think **content()** must be val with getter/setter, and **isEmpty()** must be val with getter.
Also I suggest to add **isBlank** property/

File: src/main/java/androidx/widget/TextView.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Returns the text displayed by the [TextView].
+ */
+fun TextView.content() = text.toString()
+
+/**
+ * Returns `true` if this char sequence is empty (contains no characters).
+ */
+fun TextView.isEmpty() = text.isEmpty()
```
Comment (JakeWharton) [inline src/main/java/androidx/widget/TextView.kt]: I don't find these two or `clear()` that useful.

File: src/main/java/androidx/widget/TextView.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Adds a [TextWatcher] to the list of those whose methods are called
+ * whenever this [TextView]'s text changes.
+ */
+fun TextView.textChange(
```
Comment (JakeWharton) [inline src/main/java/androidx/widget/TextView.kt]: This should be `inline` and its lambdas `crossinline` and it should return the `TextWatcher` instance to allow unregistering.

File: src/main/java/androidx/widget/TextView.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    after: (s: Editable) -> Unit = {},
+    before: (string: String, start: Int, count: Int, after: Int) -> Unit = { _, _, _, _ -> },
+    onTextChanged: (string: String, start: Int, count: Int, after: Int) -> Unit
+) =
+    addTextChangedListener(object : TextWatcher {
+        override fun afterTextChanged(s: Editable) = after.invoke(s)
+
+        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) =
+            before.invoke(s.toString(), start, count, after)
```
Comment (JakeWharton) [inline src/main/java/androidx/widget/TextView.kt]: Remove `toString()`

File: src/main/java/androidx/widget/TextView.kt [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) =
+            onTextChanged(s.toString(), start, before, count)
```
Comment (JakeWharton) [inline src/main/java/androidx/widget/TextView.kt]: Remove `toString()`

File: src/main/java/androidx/widget/TextView.kt (hunk 1 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: I will remove.

File: src/main/java/androidx/widget/TextView.kt (hunk 3 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Ok.

File: src/main/java/androidx/widget/TextView.kt (hunk 4 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Ok.

File: src/main/java/androidx/widget/TextView.kt (hunk 1 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Fixed.

File: src/main/java/androidx/widget/TextView.kt (hunk 2 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Fixed.

File: src/main/java/androidx/widget/TextView.kt (hunk 3 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Fixed.

File: src/main/java/androidx/widget/TextView.kt (hunk 4 above)
Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: Fixed.

**Comment by jaredsburrows** [comment]
@JakeWharton all fixed!

**Comment by jaredsburrows** [comment]
Anything else you want me to fix in this PR @JakeWharton? :)

**Comment by dovahkiin98** [comment]
@jaredsburrows Why did you remove isEmpty and clear()? 
if you think they're "too obvious", ViewGroup has isEmpty and size which are too obvious too

**Comment by jaredsburrows** [comment]
@dovahkiin98 @JakeWharton asked me to.

**Comment by jaredsburrows** [comment]
Can we merge? @JakeWharton? :)

**Comment by JakeWharton** [comment]
Every time you ask you get dropped to the bottom of the queue. We are aware
of the open issues and PRs and don't need reminding.

File: src/main/java/androidx/widget/TextView.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Returns `true` if this char sequence is empty (contains no characters).
+ */
+fun TextView.isEmpty() = text.isEmpty()
+
+/**
+ * Adds a [TextWatcher] to the list of those whose methods are called
+ * whenever this [TextView]'s text changes.
+ */
+inline fun TextView.textChange(
```
Comment (romainguy) [inline src/main/java/androidx/widget/TextView.kt]: We need a better name here. Maybe addTextListener() and then add doOnBeforeTextChanged/doOnAfterTextChanged() like in the Animator.kt file?

Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: I agree with the name change. Let me make the update.

Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: I renamed `testChange()` to `addTextListener()` and added both `doOnBeforeTextChanged()` and `doOnAfterTextChanged()` with tests!

File: src/main/java/androidx/widget/TextView.kt (hunk 1 above)
Comment (JakeWharton) [reply src/main/java/androidx/widget/TextView.kt]: Forget to push up changes?

Comment (jaredsburrows) [reply src/main/java/androidx/widget/TextView.kt]: I removed both `isEmpty()` and `clear()`.

**Comment by jaredsburrows** [comment]
@romainguy @JakeWharton All updated!

**Comment by Tougee** [comment]
Didn't see your updates, I update base on your commit @jaredsburrows [here](https://github.com/android/android-ktx/pull/433)

**Comment by jaredsburrows** [comment]
@romainguy I updated the build but I am not sure how this error affected my branch:
```
#!/bin/bash -eo pipefail
./gradlew check --no-daemon
:checkApi


No source files specified: recursively including all sources found in the source path
...
api/current.txt:654: error: Removed class androidx.core.widget.TextViewKt [RemovedClass:8]

warning: Added package androidx.widget [AddedPackage:2]Aborting: Found compatibility problems with --check-compatibility
:checkApi FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkApi'.
> Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 255

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 22s
1 actionable task: 1 executed
Exited with code 1
```

**Comment by jaredsburrows** [comment]
@romainguy @JakeWharton How can I get the CI `test_instrumented` to pass?

**Comment by JakeWharton** [comment]
You cannot. #141

**Comment by jaredsburrows** [comment]
@romainguy @JakeWharton Review :)

**Comment by JakeWharton** [comment]
Text watcher-based listeners have been merged in http://r.android.com/720364

**Pull request closed without merging**